### PR TITLE
SD: add etcdv3 and zookeeper support

### PR DESCRIPTION
**Pull request by wklken**
We deployed thanos 0.3.2 in K8S Cluster, k8s DNS as in-cluster-SD(for all components of thanos), and Gossip as out-cluster-SD(for all sidecars in other cluster). And we have a global etcd/zk cluster as SD servers, so we want to know if it's possible to replace `Gossip` with `etcd`.

While Thanos SD support File-SD/DNS/Gossip(will remove soon?), I am not sure if this is appropriate to make the `discovery` module to support etcd, zk and consul?

DNS and file-SD are just enough for most conditions, but I want to make sure is Thanos will support more SD options? 

More to add later:
- unittest
- etcd support
- push go mode dependency files
- docs and comments

related issue: #723 

## Changes

- add etcdv3 / zookeeper as SD options
- add tls support for etcdv3 SD

## Verification
- tested locally and  in a k8s cluster

### CODE REVIEW

**Comment by povilasv** [comment]
I think we shouldn't support this. And go the same road as Prometheus:

>Prometheus offers a variety of service discovery options for discovering scrape targets, including Kubernetes, Consul, and many others. If you need to use a service discovery system that is not currently supported, your use case may be best served by Prometheus' file-based service discovery mechanism, which enables you to list scrape targets in a JSON file (along with metadata about those targets).

You can read more about it in https://prometheus.io/docs/guides/file-sd/#use-file-based-service-discovery-to-discover-scrape-targets

File SD is more than enough to build your own SD options.

@bwplotka / @GiedriusS  what do you guys think?

**Comment by povilasv** [comment]
This also introduces a bunch of dependencies like etcd client and zk client which means more work for us to update and maintain. So I'm definetely against this. Btw if maintainers agree we should document this.

https://github.com/improbable-eng/thanos/pull/949/files#diff-d0f0531db672628e9123707a128d160cR11

**Comment by bwplotka** [comment]
Hi 🍺

Thanks for this, however it is not our goal to extend the discovery package in Thanos to other providers. That was stated in the very beginning of discovery pkg creation and reasons are the same as gossip removal decision (it will be removed soon, see: https://github.com/improbable-eng/thanos/blob/master/docs/proposals/approved/201809_gossip-removal.md)

TL;DR is that we don't want to make same mistake as Prometheus did, but allowing so many providers and fighting to maintain this with thousands of lines of unnecessary dependency hell.

However, we also know your pain, so I would propose something different - implementing this in separate repository as a separate sidecar using Thanos FileSD. We are happy to help reviewing and happy to link your solution in official Thanos docs. ((: Let me know what do you think!

**Comment by wklken** [comment]
I think I get the point. It's a good idea to separate the sd options, thanks and I will take sometime to have a try.

**Comment by bwplotka** [comment]
Additionally Thanos FileSD API is same as Prometheus FileSD, so you can use the same thing here & there (:
